cql3/statements/describe_statement.cc

```diff
@@ -450,11 +539,40 @@ future<std::vector<description>> describe_all_keyspace_elements(const data_dicti
     inserter(co_await aggregates(replica_db, ks, true));
     inserter(co_await tables(db, ks_meta, with_internals));
 
-    co_return result; 
+    co_return result;
 }
 
+future<std::vector<description>> describe_auth_and_service_levels(const auth::service& auth_service,
+        const qos::service_level_controller& sl_controller, const data_dictionary::database& db, bool with_salted_hashes)
+{
+    auto& replica_db = db.real_database();
+    const auto& role_manager = auth_service.underlying_role_manager();
+    const auto& authorizer = auth_service.underlying_authorizer();
+
+    auto role_descs = co_await generate_descriptions(replica_db, co_await role_manager.get_role_info(with_salted_hashes), true);
+    auto grant_descs = co_await generate_descriptions_with_multiple_generators(co_await role_manager.get_role_grants(), replica_db, true);
+    auto permission_descs = co_await generate_descriptions_with_multiple_generators(co_await authorizer.list_all(), replica_db, true);
+    auto attached_sl_descs = co_await generate_descriptions(replica_db, co_await role_manager.get_attached_service_levels(), true);
+
+    // We cannot allow for modifying the cache while we're accessing it, so we're copying it here.
+    // TODO: This can be dropped once we protect any changes while `DESC SCHEMA` is being performed.
+    const auto sl_cache = sl_controller.get_service_levels_cache();
```

Comment:
reffering to cache at this levels seems wrong, access should be via SL service/controller, no?